File: Dynamic Programming/70_climbing_stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [0 for _ in range(n + 1)]

        def helper(dp, n):
            if n < 0:
                return 0
            if n == 0: 
                return 1

            if dp[n] != 0:
                return dp[n]
            else:
                dp[n] = helper(dp, n - 1) + helper(dp, n - 2)

            return dp[n]

        return helper(dp, n)
    
# Revised - 01/06/2025

File: Dynamic Programming/70_climbing_stairs.cpp
class Solution {
    public:
        int climbStairs(int n) {
            int one = 1, two = 1;
            for(int i = 2; i <= n; i += 1) {
                int temp = two;
                two += one;
                one = temp;
            }
            return two;
        }
    };

// revised - 02/12/2025